salt/_states/docker_registry.py

```diff
@@ -0,0 +1,202 @@
+                                'digest': digest,
+                            },
+                        },
+                    })
+                else:
+                    ret['comment'] = 'Failed to push image'
+                    ret['result'] = False
+
+                return ret
+        else:
+            assert False, 'Logic error'
+    elif image_manifest is not None and digest is None:
+        if not force:
+            ret['comment'] = 'Image exists'
+            ret['result'] = True
+        else:
+            if __opts__['test']:
+                ret['comment'] = 'Would force-push image'
+                ret['result'] = None
+                ret['pchanges'].update({
+                    name: {
+                        'old': {
+                            'digest': image_manifest['Digest'],
+                        },
+                        'new': {
+                            'digest': digest,
+                        },
+                    },
+                })
+            else:
+                result = __salt__['docker_registry.push_docker_archive'](
+                    archive_path=archive_path,
+                    destination=name,
+                    dest_tls_verify=tls_verify)
+                if result:
+                    ret['comment'] = 'Force-pushed image'
+                    ret['result'] = True
+                    ret['changes'].update({
+                        name: {
+                            'old': {
+                                'digest': image_manifest['Digest'],
+                            },
+                            'new': {
+                                'digest': digest,
+                            },
+                        },
+                    })
+                else:
+                    ret['comment'] = 'Failed to push image'
+                    ret['result'] = False
+
+        return ret
+    elif image_manifest is None:
+        if __opts__['test']:
+            ret['comment'] = 'Would push image'
+            ret['result'] = None
+            ret['pchanges'].update({
+                name: {
+                    'old': {},
+                    'new': {
+                        'digest': digest,
+                    },
+                },
+            })
+            return ret
+
+        result = __salt__['docker_registry.push_docker_archive'](
+            archive_path=archive_path,
+            destination=name,
+            dest_tls_verify=tls_verify)
+
+        if result:
+            ret['comment'] = 'Pushed image'
+            ret['changes'].update({
+                name: {
+                    'old': {},
+                    'new': {
+                        'digest': digest,
+                    },
+                },
+            })
+            ret['result'] = True
+        else:
+            ret['comment'] = 'Failed to push image'
+            ret['result'] = False
+
+        return ret
+    else:
+        assert False, 'Logic error'
```

Comment:
This function is too long (~150 lines) with several piece of code duplicated (l.74-81 vs l.61-66)or very similar (`ret['changes'].update`).

This could probably be factored with 1-2 helper functions.

Reply:
Agree, will do so in other PR.